### ISSUE TITLE
Why there are only <unk> symbols while evaluation during training?

### ISSUE THREAD

**Issue by ssokhey** [tensorflow/nmt#150]
I am training my NMT system. I haven't applied subword units yet (just want to compare the results). I noticed I'm only getting <unk> symbols while training evaluation which leads to BLEU 0.00.

Here is a snippet:

```
eval dev: perplexity 1.15, time 1s, Fri Oct 20 22:39:36 2017.
  global step 7100 lr 1 step-time 0.63s wps 7.44K ppl 1.15 bleu 0.00
  global step 7200 lr 1 step-time 0.62s wps 7.58K ppl 1.15 bleu 0.00
  global step 7300 lr 1 step-time 0.62s wps 7.47K ppl 1.15 bleu 0.00
  global step 7400 lr 1 step-time 0.61s wps 7.61K ppl 1.15 bleu 0.00
  global step 7500 lr 1 step-time 0.61s wps 7.64K ppl 1.16 bleu 0.00
  global step 7600 lr 1 step-time 0.61s wps 7.62K ppl 1.16 bleu 0.00
  global step 7700 lr 1 step-time 0.61s wps 7.62K ppl 1.16 bleu 0.00
  global step 7800 lr 1 step-time 0.62s wps 7.63K ppl 1.15 bleu 0.00
  global step 7900 lr 1 step-time 0.62s wps 7.42K ppl 1.15 bleu 0.00
  global step 8000 lr 1 step-time 0.63s wps 7.44K ppl 1.15 bleu 0.00
# Save eval, global step 8000
2017-10-20 22:49:59.030309: W tensorflow/core/kernels/lookup_util.cc:362] Table trying to initialize from file nmt/data/vocab.hi is already initialized.
2017-10-20 22:49:59.033932: W tensorflow/core/kernels/lookup_util.cc:362] Table trying to initialize from file nmt/data/vocab.en is already initialized.
2017-10-20 22:49:59.037640: W tensorflow/core/kernels/lookup_util.cc:362] Table trying to initialize from file nmt/data/vocab.en is already initialized.
  loaded infer model parameters from nmt/nmt_attn/translate.ckpt-8000, time 1.49s
  # 372
    src: सुप्रीम कोर्ट के एक फैसले से तिलमिलाए केंद्रीय ग्रामीण विकास मंत्री जयराम रमेश ने अदालतों के लक्ष्मण रेखा लांघने को खतरनाक चलन बताया है । 
    ref: stung by the supreme court decision the rural development minister , jairam ramesh , said that the practice of exceeding their powers by the court is dangerous . 
    nmt: <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk> <unk>

2017-10-20 22:50:00.880712: W tensorflow/core/kernels/lookup_util.cc:362] Table trying to initialize from file nmt/data/vocab.en is already initialized.
2017-10-20 22:50:00.889831: W tensorflow/core/kernels/lookup_util.cc:362] Table trying to initialize from file nmt/data/vocab.hi is already initialized.
  loaded eval model parameters from nmt/nmt_attn/translate.ckpt-8000, time 1.29s
2017-10-20 22:50:02.547985: W tensorflow/core/framework/op_kernel.cc:1192] Out of range: End of sequence
	 [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?,?], [?,?], [?,?], [?], [?]], output_types=[DT_INT32, DT_INT32, DT_INT32, DT_INT32, DT_INT32], _device="/job:localhost/replica:0/task:0/device:CPU:0"](Iterator)]]
2017-10-20 22:50:02.548252: W tensorflow/core/framework/op_kernel.cc:1192] Out of range: End of sequence
	 [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?,?], [?,?], [?,?], [?], [?]], output_types=[DT_INT32, DT_INT32, DT_INT32, DT_INT32, DT_INT32], _device="/job:localhost/replica:0/task:0/device:CPU:0"](Iterator)]]
2017-10-20 22:50:02.548396: W tensorflow/core/framework/op_kernel.cc:1192] Out of range: End of sequence
	 [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?,?], [?,?], [?,?], [?], [?]], output_types=[DT_INT32, DT_INT32, DT_INT32, DT_INT32, DT_INT32], _device="/job:localhost/replica:0/task:0/device:CPU:0"](Iterator)]]

```
What could be the possible reason?

**Comment by kmario23** [comment]
This is normal since the model is evaluated on the last saved checkpoint. Two kinds of evaluation happens: 

1.  Evaluating perplexity of the training model. (given by perplexity scores, noted as `ppl`)
2.  Evaluating it on the development set to compute BLEU score.

To get a valid BLEU score (other than 0), stop training the model after some iterations (maybe 20K), and then save the latest checkpoint. Again, start training the model with the same `out_dir`, then you'll see a valid BLEU score (for your evaluation data)

**Comment by ssokhey** [comment]
@kmario23 Thanks for the insight.
But something went wrong. I trained my model with 50000 steps and after training completion, I generated translations. Every translation is a <unk> symbol.
I'm trying to figure out what went wrong.

**Comment by ssokhey** [comment]
I was creating the vocabulary in a wrong way. It was my fault.

**Comment by mahjabeen-sust** [comment]
@ssokhey how to create vocabulary then? I'm facing the same problem. Moreover my system automatically switch from GPU to CPU after this "out of range" problem occurs. Any idea why that happens?

**Comment by d2sys** [comment]
@mahjabeen-sust , you can use some useful functions from here:
https://github.com/tensorflow/models/blob/master/tutorials/rnn/translate/data_utils.py
 And you should extract the vocabulary from all your training data.
For some languages you can find an already extracted vocabulary...

**Comment by mohammedayub44** [comment]
@ssokhey Could you please briefly explain when you meant your creating the vocabulary in the wrong way. I'm experiencing the same problem with my training. 
Is it the Datatype, Should it be just be one big string with '\n' appended, if I'm correct ?

Thanks !

Mohammed Ayub

**Comment by kuab2686** [comment]
Hi all,
I have used the same model architecture and could reproduce the results for IWSLT Evaluation Campaign.
But when I used the above model with attention for english to english text summarization, I am getting only <unk> as nmt output.
I have used the vocab.en from IWSLT Evaluation Campaign. But still the same.
I used nltk.word_tokenize to generate vocab tokens, saved to excel and then
I copy pasted manually the word tokens into the vocab.en file
vocab.en has the following format and is used for both source and target
<unk>
<s>
</s>
has
gun
my


let me know what I am doing wrong?

**Comment by ssokhey** [comment]
@mahjabeen-sust @d2sys @mohammedayub44 @kuab2686 

I just saw the notifications. Please use the script at [this link](https://github.com/google/seq2seq/blob/master/bin/tools/generate_vocab.py). Make sure you remove the word counts which comes along with vocabulary (A little change in the script).

Also after generating vocabulary, add these three lines on top of every vocab file. (The architecture expects these symbols):
```
<unk>
<s>
</s>
```

**Comment by kuab2686** [comment]
update:
Cleaned all train, dev and test sets.
generated tokens from entire train set
changed learning rate to 0.01
unk no longer appearing as output

but now getting the same word 'great' as output in all cases
